uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/preferences/Preferences.kt

```diff
@@ -135,6 +141,23 @@ open class Preferences(private val storage: Storage) {
             storage.putBoolean("pref_short_toggle", enabled)
         }
 
+    internal fun setActiveNotifications(activeNotifications: Map<Habit, NotificationTray.NotificationData>) {
+        val activeById = activeNotifications.mapKeys { it.key.id }
+        val serialized = Json.encodeToString(activeById)
+        storage.putString("pref_active_notifications", serialized)
+    }
+
+    internal fun getActiveNotifications(habitList: HabitList): HashMap<Habit, NotificationTray.NotificationData> {
+        val serialized = storage.getString("pref_active_notifications", "")
+        return if (serialized == "") {
+            HashMap()
+        } else {
+            val activeById = Json.decodeFromString(MapSerializer(Long.serializer(), NotificationTray.NotificationData.serializer()), serialized)
```

Comment:
What happens if the decoding fails? Does the app fail to boot? I think this could happen if: (i) user receives some notifications; (ii) we update `NotificationData` with some new/renamed fields; (iii) user installs updates and restarts the app. The new version of the app would not be able to read serialized data from the old version.

I suggest adding a try/catch block here, and returning `HashMap()` if the operation fails.

Reply:
Yes currently there would be an uncaught exception if decoding fails, and the app would crash on startup. That is of course bad, even though unlikely. Returning an empty map in case it cannot be deserialized is fine, as (as far as I can see) no other functionality depends on all active notifications being available in the map.

What we could do is show a notification that notifications could not be restored and the user should make sure to look at the habits manually, but regarding that it is a corner case it's probably not worth it.

Btw., decoding will not fail if fields not present in the encoded string have default values in the Kotlin class.